I590 - Applied Data Science

Wrangling using Python

Created by Vivek Vijayaraghavan

Date: 04/04/2019

Objective: Code Portfolio for wrangling using python

<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Data-Wrangling-using-Python" data-toc-modified-id="Data-Wrangling-using-Python-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Data Wrangling using Python</a></span><ul class="toc-item"><li><span><a href="#Overview" data-toc-modified-id="Overview-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Overview</a></span></li><li><span><a href="#Pre-Requisites" data-toc-modified-id="Pre-Requisites-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>Pre-Requisites</a></span></li><li><span><a href="#References" data-toc-modified-id="References-1.3"><span class="toc-item-num">1.3&nbsp;&nbsp;</span>References</a></span></li></ul></li><li><span><a href="#Data-exploration-with-Pandas" data-toc-modified-id="Data-exploration-with-Pandas-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Data exploration with Pandas</a></span><ul class="toc-item"><li><span><a href="#Inspecting-the-data" data-toc-modified-id="Inspecting-the-data-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Inspecting the data</a></span></li><li><span><a href="#Indexing-&amp;-Selecting-data" data-toc-modified-id="Indexing-&amp;-Selecting-data-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Indexing &amp; Selecting data</a></span></li><li><span><a href="#Split---Apply---Combine" data-toc-modified-id="Split---Apply---Combine-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Split - Apply - Combine</a></span></li><li><span><a href="#Data-Transformation" data-toc-modified-id="Data-Transformation-2.4"><span class="toc-item-num">2.4&nbsp;&nbsp;</span>Data Transformation</a></span></li></ul></li></ul></div>

# Data Wrangling using Python

## Overview

## Pre-Requisites
Load the relevant libraries and set options

In [1]:
# Load the libraries that you will require in this notebook.
from os import path
import numpy as np
import pandas as pd
from matplotlib.pyplot import figure, show, rc
import matplotlib.pyplot as plt
import matplotlib.mlab as mlab
import json

# Set plotting options.
%pylab inline
pylab.rcParams['figure.figsize'] = (10, 8)

Populating the interactive namespace from numpy and matplotlib


## References

# Data exploration with Pandas

This section introduces other tools for pre-processing and exploring data in Python using Pandas. For this purpose, a data set that was collected using the StudentLife app, built by researchers from Dartmouth College, will be used. The Android app monitored readings from smartphone sensors carried by 48 Dartmouth students during a 10-week term. This was in order to assess their mental health (depression, loneliness, and stress), academic performance (grades across all their classes, term GPA, and cumulative GPA), and behavioral trends (how factors such as stress, sleep, and visits to the gym, change in response to college workload – assignments, midterms, and finals – as the term progressed).

## Inspecting the data

In [2]:
# Load a specified file to a Pandas dataframe.
df_gps = pd.read_csv('./data/gps_u02.csv')

In [3]:
# Review the data with "head()".
# gps.tail() will return the records from the end of the dataset.
df_gps.head()

time provider network_type  accuracy   latitude  longitude  altitude  \
0  1364357797  network         wifi      20.0  43.707127 -72.293314       0.0   
1  1364358997  network         wifi      20.0  43.707123 -72.293319       0.0   
2  1364360197  network         wifi      20.0  43.707166 -72.293272       0.0   
3  1364361397  network         wifi      20.0  43.707158 -72.293285       0.0   
4  1364362597  network         wifi      20.0  43.707131 -72.293320       0.0   

   bearing  speed travelstate  
0      0.0    0.0  stationary  
1      0.0    0.0  stationary  
2      0.0    0.0  stationary  
3      0.0    0.0  stationary  
4      0.0    0.0  stationary

In [4]:
# Review the data with "tail()".
df_gps.tail()

time provider network_type  accuracy   latitude  longitude  \
6602  1370054701  network         wifi    60.935  43.707113 -72.293186   
6603  1370055301  network         wifi    59.683  43.707161 -72.293165   
6604  1370055901  network         wifi    60.811  43.707155 -72.293197   
6605  1370056501  network         wifi    63.253  43.707124 -72.293179   
6606  1370057094  network         wifi    59.327  43.707131 -72.293187   

      altitude  bearing  speed travelstate  
6602       0.0      0.0    0.0  stationary  
6603       0.0      0.0    0.0  stationary  
6604       0.0      0.0    0.0  stationary  
6605       0.0      0.0    0.0  stationary  
6606       0.0      0.0    0.0  stationary

In [5]:
# You can check the number of cases, as well as the number of fields, using the shape method, as shown below.
df_gps.shape

(6607, 10)

In the result displayed, you can see the data has 6,607 records, each with 10 columns.

The “info()” and “describe()” methods can also be used to provide additional information about the data set. The “info()” method provides a concise summary of the data; from the output, it provides the type of data in each column, the number of non-null values in each column, and how much memory the data frame is using.

In [6]:
# Review data types with "info()".
df_gps.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6607 entries, 0 to 6606
Data columns (total 10 columns):
time            6607 non-null int64
provider        6607 non-null object
network_type    6344 non-null object
accuracy        6607 non-null float64
latitude        6607 non-null float64
longitude       6607 non-null float64
altitude        6607 non-null float64
bearing         6607 non-null float64
speed           6607 non-null float64
travelstate     6290 non-null object
dtypes: float64(6), int64(1), object(3)
memory usage: 516.2+ KB


The “describe()” DataFrame method returns summary statistics for a quick and simple description of the data. These include count, mean, median, mode, minimum value, maximum value, range, and standard deviation.

In [7]:
# Review numeric data with "describe()".
df_gps.describe()

time     accuracy     latitude    longitude     altitude  \
count  6.607000e+03  6607.000000  6607.000000  6607.000000  6607.000000   
mean   1.367687e+09    74.158883    43.706182   -72.292437     5.554109   
std    1.624653e+06   255.141546     0.025016     0.017621    27.636964   
min    1.364358e+09     3.000000    43.512803   -72.754270     0.000000   
25%    1.366349e+09    25.118500    43.706573   -72.293224     0.000000   
50%    1.368036e+09    36.122000    43.707108   -72.293184     0.000000   
75%    1.369049e+09    58.230000    43.707131   -72.288852     0.000000   
max    1.370057e+09  4027.000000    44.353047   -72.267886   306.100006   

           bearing        speed  
count  6607.000000  6607.000000  
mean      2.390404     0.032516  
std      23.205340     0.625369  
min       0.000000     0.000000  
25%       0.000000     0.000000  
50%       0.000000     0.000000  
75%       0.000000     0.000000  
max     351.700000    35.003570

## Indexing & Selecting data

Data loaded using Pandas is automatically appended with metadata for axis labeling, which does the following:
 - Identifies data (i.e., provides metadata) using known indicators, which is important for analysis, visualization, and interactive console display.
 - Enables automatic and explicit data alignment.
 - Allows intuitive getting and setting of subsets of the data set.

The following examples demonstrate how to slice, dice, and generally get and set subsets from pandas objects.

The “.loc” attribute is the primary access method. The following values are valid inputs:
 - A single label (e.g., 5 or 'a'). (Note that 5 is interpreted as a label of the index. This use is not an integer position along the index.)
 - A list or array of labels ['a', 'b', 'c'].
 - A slice object with labels 'a':'f'. (Note that contrary to usual python slices, both the start and the stop are included.)
 - A boolean array.

In [8]:
# Select the first 10 rows.
df_gps.loc[0:9]

time provider network_type  accuracy   latitude  longitude  altitude  \
0  1364357797  network         wifi    20.000  43.707127 -72.293314       0.0   
1  1364358997  network         wifi    20.000  43.707123 -72.293319       0.0   
2  1364360197  network         wifi    20.000  43.707166 -72.293272       0.0   
3  1364361397  network         wifi    20.000  43.707158 -72.293285       0.0   
4  1364362597  network         wifi    20.000  43.707131 -72.293320       0.0   
5  1364363797  network         wifi    20.000  43.707123 -72.293288       0.0   
6  1364365002      gps          NaN     8.000  43.707182 -72.292976     159.5   
7  1364366197  network         wifi    31.201  43.707122 -72.293282       0.0   
8  1364367397  network         wifi    20.000  43.707134 -72.293293       0.0   
9  1364368597  network         wifi    29.711  43.707122 -72.293256       0.0   

   bearing  speed travelstate  
0      0.0    0.0  stationary  
1      0.0    0.0  stationary  
2      0.0    0.0  stationary  
3      0.0    0.0  stationary  
4      0.0    0.0  stationary  
5      0.0    0.0  stationary  
6      0.0    0.0         NaN  
7      0.0    0.0  stationary  
8      0.0    0.0  stationary  
9      0.0    0.0  stationary

In [9]:
# Select the columns ('provider','latitude','longitude') for the rows labelled 10,11, ...,15.
df_gps.loc[10:15, ['provider','latitude','longitude']]

provider   latitude  longitude
10  network  43.707163 -72.293251
11  network  43.707155 -72.293281
12  network  43.707156 -72.293238
13  network  43.707152 -72.293245
14  network  43.707128 -72.293309
15  network  43.707159 -72.293255

In [10]:
# Access records where the provider is 'gps'. Return the first 5 rows from the filtered data frame.
df_gps.loc[df_gps['provider']=='gps', :].head()

time provider network_type  accuracy   latitude  longitude  \
6    1364365002      gps          NaN       8.0  43.707182 -72.292976   
21   1364382891      gps          NaN      13.0  43.707372 -72.293165   
125  1364507801      gps          NaN      10.0  43.706560 -72.292041   
171  1364562996      gps          NaN      19.0  43.707121 -72.293339   
173  1364565402      gps          NaN      10.0  43.707235 -72.285593   

       altitude  bearing  speed travelstate  
6    159.500000      0.0   0.00         NaN  
21   144.899994      0.0   0.00         NaN  
125  138.199997    148.7   1.25         NaN  
171  119.199997      0.0   0.00         NaN  
173  150.100006     33.2   1.75         NaN

In [11]:
# Select the first two rows using intger-based indexing.
df_gps.iloc[0:2,:]

time provider network_type  accuracy   latitude  longitude  altitude  \
0  1364357797  network         wifi      20.0  43.707127 -72.293314       0.0   
1  1364358997  network         wifi      20.0  43.707123 -72.293319       0.0   

   bearing  speed travelstate  
0      0.0    0.0  stationary  
1      0.0    0.0  stationary

In [12]:
# Select the first 5 rows and columns numbered 2, 3, and 4. 
df_gps.iloc[0:5,2:5]

network_type  accuracy   latitude
0         wifi      20.0  43.707127
1         wifi      20.0  43.707123
2         wifi      20.0  43.707166
3         wifi      20.0  43.707158
4         wifi      20.0  43.707131

## Split - Apply - Combine

When approaching a data analysis problem, it is a good idea to break it into manageable pieces, perform some operations on each of the pieces, and then put everything back together again. The Pandas “groupby” method draws largely from the split-apply-combine strategy for data analysis. For example, say you are interested in splitting your data depending on the provider type. The following example illustrates the use of “groupby()” on your DataFrame.

In [13]:
grp = df_gps.groupby(['provider'])
print(grp)

In [14]:
# Calling "count" on the grouped object returns the total number of non-null values within each column.
grp.count()

time  network_type  accuracy  latitude  longitude  altitude  \
provider                                                                
gps        263             0       263       263        263       263   
network   6344          6344      6344      6344       6344      6344   

          bearing  speed  travelstate  
provider                               
gps           263    263            0  
network      6344   6344         6290

In [15]:
# To get the total number of records in each group, you can use the "size".
grp.size()

provider
gps         263
network    6344
dtype: int64

In [16]:
# If we are interested in the average accuracy for each of the groups, we can do the following:
grp.mean()['accuracy']

provider
gps        13.026616
network    76.693213
Name: accuracy, dtype: float64

## Data Transformation

Sometimes you are provided with data that needs to be transformed into a form that is more familiar, so that you can do other analyses, or search for patterns or other insights. DataFrames in Pandas are equipped with an "apply" method that can help with such transformation. These transformations may require the use of other functions. For example, the time field in your GPS data is in a form called Unix time (also known as POSIX time or Epoch time), which is a system for describing instants in time, defined as the number of seconds that have elapsed since 00:00:00 Coordinated Universal Time (UTC), Thursday, 1 January 1970, not counting leap seconds. Unix time is a single signed integer number that increments every second, without requiring the calculations to determine year, month, day of month, hour, and minute, necessary for intelligibility to humans.

The Pandas “to_datetime” method will be used to convert the time into something that makes sense to you, by introducing another Python style construct called “lambda” functions inside an apply method. Python supports the creation of anonymous functions (i.e., functions that are not bound to a name) at runtime, using a construct called "lambda". While using lambda expressions is never a necessity, they allow you to perform quick hacks without having to bloat your code base with too many rarely-used functions. Let’s now demonstrate using the apply method.

In [17]:
# Intrspect the top rows of the dataframe
df_gps.head()

time provider network_type  accuracy   latitude  longitude  altitude  \
0  1364357797  network         wifi      20.0  43.707127 -72.293314       0.0   
1  1364358997  network         wifi      20.0  43.707123 -72.293319       0.0   
2  1364360197  network         wifi      20.0  43.707166 -72.293272       0.0   
3  1364361397  network         wifi      20.0  43.707158 -72.293285       0.0   
4  1364362597  network         wifi      20.0  43.707131 -72.293320       0.0   

   bearing  speed travelstate  
0      0.0    0.0  stationary  
1      0.0    0.0  stationary  
2      0.0    0.0  stationary  
3      0.0    0.0  stationary  
4      0.0    0.0  stationary

In [18]:
# Transform the data by adding an additional column
df_gps.loc[:,['time']] = df_gps.time.apply(lambda x: pd.to_datetime(x,unit='s'))

In [19]:
# Review the data, including the transformed column.
df_gps.head()

time provider network_type  accuracy   latitude  longitude  \
0 2013-03-27 04:16:37  network         wifi      20.0  43.707127 -72.293314   
1 2013-03-27 04:36:37  network         wifi      20.0  43.707123 -72.293319   
2 2013-03-27 04:56:37  network         wifi      20.0  43.707166 -72.293272   
3 2013-03-27 05:16:37  network         wifi      20.0  43.707158 -72.293285   
4 2013-03-27 05:36:37  network         wifi      20.0  43.707131 -72.293320   

   altitude  bearing  speed travelstate  
0       0.0      0.0    0.0  stationary  
1       0.0      0.0    0.0  stationary  
2       0.0      0.0    0.0  stationary  
3       0.0      0.0    0.0  stationary  
4       0.0      0.0    0.0  stationary